### ISSUE TITLE
Accessing MongoDB/nosql databases

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/docs#3350]
Check with SME for how we should handle MongoDB relative to the other JDBC/database info in the OL docs. eg should we have an "Accessing nosql databases" topic that covers MongoDB, CouchDB, and Cloudant? What's best practice: using CDI or using cloudant feature (I think the former?)? If you can use CDI for any nosql db, we should just document that and ignore the cloudant feature, unless there are good reasons that someone might use the cloudant approach instead (in which case, clearly state what they are).

There's a blog post about [how to access a MongoDB database from an OL app](https://openliberty.io/blog/2019/02/19/mongodb-with-open-liberty.html?_ga=2.207768594.1663611092.1606818058-1399812591.1606212512) by @mswatosh . He's probably best person as SME for this. Talk to him about how this should fit in with the database info we've already published in the OL docs.

If he agrees this is useful as a topic, make sure not to talk about versions and how you used to do this with the mongodb feature - that's irrelevant here. Instead, just talk about how to access MongoDB databases from OL (which is by using CDI as he describes in his post).

But, where it talks about multiple apps and shared libraries, check that this fits with containers and cloud context.

In case any of it's useful, here are the KC topics:

There are MongoDB topics in the KC that do stuff in the old way. Not sure if any of that is useful: https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_mongodb_create.html
Couchdb KC topics: https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_couchdb.html

There are also CouchDB topics in the KC: https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_couchdb.html

And how to use the Cloudant feature (instead of CDI): https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_couchdb_config.html (but check whether we should doc this or if we should just encourage users to take the CDI approach). Also, this Cloudant topic: https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_couchdb_configdriver.html seems to just be talking about the elements and attributes that are documented in the [gendoc](https://www.openliberty.io/docs/20.0.0.12/reference/config/cloudant.html). So if this is mentioned in the OL doc, it should just be some sentences with other info and then link to the gendoc (not a whole separate topic like this).

**Comment by mswatosh** [comment]
There's also a close to being published guide here for mongodb that may have additional useful information:
https://github.com/OpenLiberty/draft-guide-mongodb

**Comment by ManasiGandhi** [comment]
# Preliminary Content Design

- **Who is the audience?** Developers, administrators
- **How does this topic fit in the docs?** Concept topic.
- **Consistency with existing docs and other docs in progress** (Cloudant integration feature, Validating server connections
- **What OL-specific information does this doc cover?** How does it fit w/ OL strategy (cloud-native, containerized- Accessing nosql databases with Open Liberty
- **Who is the SME**- Waiting on Fred Rowe to get back on this.
- **What are existing resources**

**Blog post**- https://openliberty.io/blog/2019/02/19/mongodb-with-open-liberty.html?_ga=2.207768594.1663611092.1606818058-1399812591.1606212512
**Guide**- https://github.com/OpenLiberty/draft-guide-mongodb
**KC topics**
https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_mongodb_create.html
https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_couchdb.html
https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_couchdb_configdriver.html

**Comment by ManasiGandhi** [comment]
@mswatosh here is a content proposal for accessing NoSQL databases with Open Liberty.

# Content proposal for accessing NoSQL databases with Open Liberty

You can access a NoSQL database with Open Liberty without enabling any feature in the server.xml file. You can configure access to a NoSQL database with a CDI producer.

## Access MongoDB with CDI

Using MongoDB with Open Liberty previously meant enabling the mongodb-2.0 feature and configuring several elements in the `server.xml` file, but was limited to certain versions of MongoDB. Now, with improvements in CDI and the introduction of MicroProfile Config, you can easily configure access to any version of MongoDB with a CDI producer.

## A CDI producer for MongoDB

With a CDI producer, you are no longer restricted to MongoDB 2.X versions, and can easily provide a MongoDatabase to your application. This example demonstrates how to create a CDI producer to inject a MongoDatabase:

```
@ApplicationScoped
public class MongoProducer {

    @Produces
    public MongoClient createMongo() {
        return new MongoClient(new ServerAddress(), new MongoClientOptions.Builder().build());
    }

    @Produces
    public MongoDatabase createDB(MongoClient client) {
        return client.getDatabase("testdb");
    }

    public void close(@Disposes MongoClient toClose) {
        toClose.close();
    }
}
```
You can also use a CDI producer to inject a MongoDatabase in a JAX-RS application, include authentication, and configure the MongoDB driver. For more information, see [Access any version of MongoDB with Open Liberty using CDI](https://openliberty.io/blog/2019/02/19/mongodb-with-open-liberty.html?_ga=2.207768594.1663611092.1606818058-1399812591.1606212512).

## Access CouchDB with CDI

For accessing CouchDB with CDI in Open Liberty, you can use Cloudant as the CouchDB client, instead of the Mongoclient.

For more information, see the[ Cloudant integration feature](https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.7/reference/feature/cloudant-1.0.html).

**Comment by ManasiGandhi** [comment]
Slack comments from Tracy Burroughs

- We recommend customers either use the cloudant-1.0 feature, or follow the instruction in the blog post mentioned in your issue  about how to access a MongoDB database from an OL app by @mswatosh; and basically just use a couchdb client instead of a mongodb client.
- Focus on providing information about just MongoDB and Cloudant (since Cloudant is the IBM supplied CouchDB client).  
- The blog post by @mswatosh is definitely the way to go for MongoDB; not sure about Cloudant.  
- Check with Zombies (Fred's team) on whether they still recommend using the cloudant-1.0 feature, or would instead recommend using Cloudant via CDI as described by @mswatosh. If they recommend using CDI, then perhaps additional documentation could be provided similar to the @mswatosh blog post for Mongodb.

**Comment by ManasiGandhi** [comment]
Slacked with Nathan Rauh and he agreed that for CouchDB, Cloudant can be used as the CouchDB client via CDI, instead of the Mongoclient.

**Comment by mswatosh** [comment]
SME review comments:

- I would remove any mention of the mongodb-2.0 feature, and the mention of it being limited to certain MongoDB versions. The 2.X versions that are mentioned haven't been supported for over a year at this point. Maybe just include a note at the end that the mongodb-2.0 feature is stabilized, and not compatible with supported MongoDD versions.

- I would include showing how to Inject the MongoDatabase as well, and instead of JAX-RS application, just say application since this isn't restricted to JAX-RS.

- Access CouchDB with CDI may need some additional clarification, since the "For more information" link isn't related the CDI part before it.

**Comment by ManasiGandhi** [comment]
@mswatosh I worked on your review. Here is the draft link:

https://draft-openlibertyio.mybluemix.net/docs/21.0.0.7/access-nosql-databases.html

- [x] I would remove any mention of the mongodb-2.0 feature, and the mention of it being limited to certain MongoDB versions. The 2.X versions that are mentioned haven't been supported for over a year at this point. Maybe just include a note at the end that the mongodb-2.0 feature is stabilized, and not compatible with supported MongoDD versions. **I did not mention that the mongodb-2.0 feature is stabilized, and not compatible with supported MongoDB versions. I removed any version specific content.**

- [x] I would include showing how to Inject the MongoDatabase as well, and instead of JAX-RS application, just say application since this isn't restricted to JAX-RS.

- [x] Access CouchDB with CDI may need some additional clarification, since the "For more information" link isn't related the CDI part before it. **Removed the link as it is not relevant to the previous sentence.**

- [x] I would remove any mention of the mongodb-2.0 feature, and the mention of it being limited to certain MongoDB versions. The 2.X versions that are mentioned haven't been supported for over a year at this point. Maybe just include a note at the end that the mongodb-2.0 feature is stabilized, and not compatible with supported MongoDD versions. **I did not mention that the mongodb-2.0 feature is stabilized, and not compatible with supported MongoDB versions. I removed any version specific content.**

- [x] I would include showing how to Inject the MongoDatabase as well, and instead of JAX-RS application, just say application since this isn't restricted to JAX-RS.

- [x] Access CouchDB with CDI may need some additional clarification, since the "For more information" link isn't related the CDI part before it. **Removed the link as it is not relevant to the previous sentence.**

**Comment by ManasiGandhi** [comment]
@mswatosh Here's the updated link with changes https://draft-openlibertyio.mybluemix.net/docs/21.0.0.7/access-nosql-databases.html. I got it reviewed from Tracy Burroughs and Nathan Rauh for CouchDB and Cloudant.

**Comment by dmuelle** [comment]
# Content Review

Hi Manasi- I dont think the current draft contains enough information. It doesn't need to be exhaustive but each section should contain more context to explain what is being described. I think its fine for a blog post not to go into depth but a docs topic should provide at least some background/context to orient the reader. 

Also, avoid using a gerund in the title as this is not a task topic

## Intro 

Some of the questions the intro could answer (in a sentence or two each):
- what is a NoSQL database and why might you need to connect to one (use case)
- introduce MongoDB and CouchDB, with links to their doc
- Although. I agree that we shouldn't focus on version-specific feature information, should mention that earlier versionns used the mongo db and couch db features, which are now stabilized. To connect to newer versions of either database, you can configure a CDI producer in you application code.
- what is a CDI producer, link to the [Injecting Dependencies into Microservices guide](https://openliberty.io/guides/cdi-intro.html) for more info
- provide a link to the [Mongo DB guide](https://www.openliberty.io/docs/21.0.0.6/reference/feature/mongodb-2.0.html) for users who want a walk-through of persisting data w/ Mongo DB. This topic focuses only on accessing the DB, whereas the guide goes into more detail for using it. You might review the guide to add more detail to the code example expaklnations in this topic.

## Access MongoDB with CDI

- throughout the topic: "MongoDatabase" ---> "MongoDatabase instance"
- there should be some explanation of what is happening in each code example- not line-by-line, but just explaining what the MongoClient and the MongoDatabase are. Refer to the guide for details
- sentence before the second example should end in a colon instead of a comma
- I think we should include the Mong DB MP config info and example in this topic instead of linking out to the blog post. Ideally, this topic should include all the info from the post so it is in one place and can be updated as needed. I think we should only link to the blog post in the see also section. This topic should contain all that info so that support could point users here instead of the blog post, as the post might not be updated if anything changes in the future, but this doc will be (hopefully).

## Access CouchDB with CDI
This section should explain what cloudant is and include an example, similar to the previous, of creating a CDI producer to inject cloudant. We don't provide this info in the blog post or guide so would add value here.

**Comment by ManasiGandhi** [comment]
@dmuelle I worked on your review. Here's the updated link https://draft-openlibertyio.mybluemix.net/docs/21.0.0.9/access-nosql-databases.html
## Intro
Some of the questions the intro could answer (in a sentence or two each):
- [x] what is a NoSQL database and why might you need to connect to one (use case)
- [x]  introduce MongoDB and CouchDB, with links to their doc
- [x]  Although. I agree that we shouldn't focus on version-specific feature information, should mention that earlier versionns used the mongo db and couch db features, which are now stabilized. To connect to newer versions of either database, you can configure a CDI producer in you application code.
- [x]  what is a CDI producer, link to the [Injecting Dependencies into Microservices guide](https://openliberty.io/guides/cdi-intro.html) for more info
- [x]  provide a link to the [Mongo DB guide](https://www.openliberty.io/docs/21.0.0.6/reference/feature/mongodb-2.0.html) for users who want a walk-through of persisting data w/ Mongo DB. This topic focuses only on accessing the DB, whereas the guide goes into more detail for using it. You might review the guide to add more detail to the code example expaklnations in this topic.
## Access MongoDB with CDI
- [x]  throughout the topic: "MongoDatabase" ---> "MongoDatabase instance"
- [x]  there should be some explanation of what is happening in each code example- not line-by-line, but just explaining what the MongoClient and the MongoDatabase are. Refer to the guide for details
- [x]  sentence before the second example should end in a colon instead of a comma
- [x]  I think we should include the Mong DB MP config info and example in this topic instead of linking out to the blog post. Ideally, this topic should include all the info from the post so it is in one place and can be updated as needed. I think we should only link to the blog post in the see also section. This topic should contain all that info so that support could point users here instead of the blog post, as the post might not be updated if anything changes in the future, but this doc will be (hopefully).
## Access CouchDB with CDI
- [ ] This section should explain what cloudant is and include an example, similar to the previous, of creating a CDI producer to inject cloudant. We don't provide this info in the blog post or guide so would add value here. **I spoke to the SMEs and as of now there is no documented example available for CouchDB with CDI similar to MongoDB**

**Comment by dmuelle** [comment]
Thanks Manasi, those changes look good. I found a few more things to address:

- [ ] NoSQL, or a 'not only SQL', database provides flexible schemas for potentially large amounts of structured, semi-structured, and unstructured data for faster development
-->
Not only SQL (NoSQL) databases provides flexible schemas for potentially large amounts of structured, semi-structured, and unstructured data.

- [ ] Earlier, using MongoDB and CouchDB required enabling the MongoDBb and CouchDBb features, which are now stabilized. To connect to newer versions of either database, you can configure a CDI producer in you application code.
--->
Before Open Liberty version 19.0.0.6, using MongoDB or Couch DB databases required you to enable the MongoDBb or CouchDBb feature. The Mongo DB feature was stabilized in version 19.0.0.4 and the CouchDB feature was stabilized in version 19.0.0.6. To connect to these databases in version 19.0.0.6 and later, you can configure a CDI producer in your application code.

- [ ] throughout, MongoDatabase--->`MongoDatabase `

- [ ] The createMongo() producer method returns an instance of MongoClient. The createDB() producer method returns an instance of MongoDatabase that depends on the MongoClient. The close()` method is a clean-up function for the MongoClient that closes the connection to the MongoDatabase instance.
---->
The `createMongo()` producer method returns a `MongoClient` instance. The `createDB()` producer method returns an \ `MongoDatabase` instance  that depends on the `MongoClient` instance. The `close()`` method is a clean-up function for the `MongoClient` instance that closes the connection to the `MongoDatabase` instance.

- [ ] user name---> username

- [ ] `You need to add the passwordUtilities-1.0 feature, available as a Maven dependency:`
This step is inconsistent with how we usually document features in the docs. Unless the feature is a special case, i don't think we need to include this step. Furthermore, it's already out of date as it specifies version 18.0.0.4, which is almost 2 years old. Check with the SME but if this feature would already be available for someone using the All Ga features package, we don't need to doc adding it from Maven. Should just tell them to enable it in server.xml. But do confirm that I am right about that :)
Also, use the text name of the feature and linbk with the feature macro on first mention

- [ ] `Further, you can enable SSL. The JSSEHelper gets the SSLContext to create a new MongoClient instance and change the createMongo method that looks like the following:`
Not sure what the intention is but it sounds like the JSSEHelper  writes the method, which I don't think is correct. Also need to refer to TLS rather than SSL and identify `JSSEHelper` and `SSLContext` as classes . 
the blog source says "You can go further and enable SSL, changing the createMongo method to look something like this:", which makes it more clear that you must rewrite the method. SO Iwould keep the level of detail that you have but reprhase to add clarity

- [ ] If there are no examples for CouchDB, I don't hink it needs a separate section. We can just move the one sentence to be after the particular example that it refers to.

**Comment by chirp1** [comment]
@dmuelle @ManasiGandhi 

Hi David and Manasi, 
I'm doing some spot checking and see a few more things beyond David's comments that I'd like to see changed. David, please work with Manasi on these comments. When Manasi has come up with the wording that the two of you agree on, I'd then like Manasi to slack me and David so that I can then review the wording.
- "without enabling any feature" doesn't seem quite right. Review the resources that Laura listed and reword/correct.
- The opening paragraph should mention CDI, but it doesn't.
- The "You can configure variables like the user and password " wording and the "To add configuration" wording are confusing, oddly worded. Rewrite the paragraph to make it clear.

**Comment by ManasiGandhi** [comment]
@chirp1 I'm still working on this issue with @dmuelle . Will work on your comments.

**Comment by chirp1** [comment]
@ManasiGandhi Hi Manasi, I don't think I see your PRs for the issue above. Would you connect your PRs for this issue to this issue with the format of # sign followed by the PR? Thanks!

**Comment by ManasiGandhi** [comment]
@chirp1 Sure Karen.

**Comment by ManasiGandhi** [comment]
@dmuelle I worked on your review. Here's the link to the updated draft https://draft-openlibertyio.mybluemix.net/docs/21.0.0.9/access-nosql-databases.html

- [x] NoSQL, or a 'not only SQL', database provides flexible schemas for potentially large amounts of structured, semi-structured, and unstructured data for faster development
-->
Not only SQL (NoSQL) databases provides flexible schemas for potentially large amounts of structured, semi-structured, and unstructured data.

- [x] Earlier, using MongoDB and CouchDB required enabling the MongoDBb and CouchDBb features, which are now stabilized. To connect to newer versions of either database, you can configure a CDI producer in you application code.
--->
Before Open Liberty version 19.0.0.6, using MongoDB or Couch DB databases required you to enable the MongoDBb or CouchDBb feature. The Mongo DB feature was stabilized in version 19.0.0.4 and the CouchDB feature was stabilized in version 19.0.0.6. To connect to these databases in version 19.0.0.6 and later, you can configure a CDI producer in your application code.

- [x] throughout, MongoDatabase--->`MongoDatabase `

- [x] The createMongo() producer method returns an instance of MongoClient. The createDB() producer method returns an instance of MongoDatabase that depends on the MongoClient. The close()` method is a clean-up function for the MongoClient that closes the connection to the MongoDatabase instance.
---->
The `createMongo()` producer method returns a `MongoClient` instance. The `createDB()` producer method returns an \ `MongoDatabase` instance  that depends on the `MongoClient` instance. The `close()`` method is a clean-up function for the `MongoClient` instance that closes the connection to the `MongoDatabase` instance.

- [x] user name---> username

- [x] `You need to add the passwordUtilities-1.0 feature, available as a Maven dependency:`
This step is inconsistent with how we usually document features in the docs. Unless the feature is a special case, i don't think we need to include this step. Furthermore, it's already out of date as it specifies version 18.0.0.4, which is almost 2 years old. Check with the SME but if this feature would already be available for someone using the All Ga features package, we don't need to doc adding it from Maven. Should just tell them to enable it in server.xml. But do confirm that I am right about that :)
Also, use the text name of the feature and linbk with the feature macro on first mention

- [x] `Further, you can enable SSL. The JSSEHelper gets the SSLContext to create a new MongoClient instance and change the createMongo method that looks like the following:`
Not sure what the intention is but it sounds like the JSSEHelper  writes the method, which I don't think is correct. Also need to refer to TLS rather than SSL and identify `JSSEHelper` and `SSLContext` as classes . 
the blog source says "You can go further and enable SSL, changing the createMongo method to look something like this:", which makes it more clear that you must rewrite the method. SO Iwould keep the level of detail that you have but reprhase to add clarity

- [x] If there are no examples for CouchDB, I don't hink it needs a separate section. We can just move the one sentence to be after the particular example that it refers to.

**Comment by dmuelle** [comment]
Fixes look good! A few more suggestions:

- [ ] `You can use the CDI producer that you created to inject a MongoDatabase instance in the application:`

It might be unclear that this sentence introduces the next example, rather than explaining the first. I suggest putting it on its own line and adding "as shown in the following example:"

- [ ] in the blog post, the security examples are under a subheading "Enhancing the CDI producer". And this sentence is included: `One of the advantages of using a CDI producer is that it can be tailored to your needs. For improved security, the createMongo method can be enhanced to include authentication with a user name and an encoded password`:

I think some version of this sentence, and a new section (maybe "Secure your CDI producer" or similar) would add some context keep the examples from running together, making them easier to parse


- [ ] `You need to enable the Password Utilities feature in the server.xml file:`

Why? Does this apply to using a password or to adding TLS? It's not clear

- [ ] `To access CouchDB with CDI, you can replace the MongoClient in the example...`
--->
`To access CouchDB with CDI, you can replace the MongoClient in the previous examples...`

- [ ] `add the following to your CDI producer:`
--->
`add the following code to your CDI producer:`

- [ ] `For more information, see Access any version of MongoDB with Open Liberty using CDI.`

recommend removing this sentence and adding the link to the "See also" section as `Blog: Access any version of MongoDB with Open Liberty using CDI`

**Comment by chirp1** [comment]
@ManasiGandhi @dmuelle 

Hi, Manasi, Thanks for making the updates. I looked over them and have comments on your updates.

- Previously I said "without enabling any feature" doesn't seem quite right. I also said to review the resources that Laura listed and reword/correct. I see that you removed the phrase instead of rewording/correcting it. I think, though, that a corrected phrase will add value to the short description and help add the essence of what the topic is about. To me, one way to interpret "You can access a NoSQL database with Open Liberty without enabling any feature in the server.xml file." is that the server.xml files does not contain a single feature in it. What isn't needed? Some features. To add clarity to the short description, figure out what those features are and include them in the short description in a meaningful way. 
 -   The opening paragraph now mentions CDI, but the new sentence repeats the title of the topic. The short description documentation says not to repeat the title. (See info about writing short descriptions on the peer review cheat sheet link.) I think that when you get the features figured out that I mentioned in the previous bullet, you can combine that info with the CDI reference to have a more meaningful short description in two sentences and about 50 words.
- I mentioned that the "You can configure variables like the user and password " wording and the "To add configuration" wording are confusing, oddly worded. I asked that you rewrite the information because the info that you were trying to convey is meaningful to the user because it would describe some of what is going on in the code. However, I see that you removed the "You can configure variables like the user and password "wording. I suggest that you add back in wording about the variables to give more context to the user. Your "You can configure variables like the user and password " wording seems to indicate that "user" and "password" are variables, but I don't think they are.

**Comment by ManasiGandhi** [comment]
@chirp1 Thanks for the review. The draft is still in the working stage. I will work on your comments in my draft.

**Comment by ManasiGandhi** [comment]
@dmuelle I worked on your review. Here's the link to the updated doc https://draft-openlibertyio.mybluemix.net/docs/21.0.0.9/access-nosql-databases.html.

- [x] `You can use the CDI producer that you created to inject a MongoDatabase instance in the application:`

It might be unclear that this sentence introduces the next example, rather than explaining the first. I suggest putting it on its own line and adding "as shown in the following example:" **Updates per our discussion in slack yesterday.**

- [x] in the blog post, the security examples are under a subheading "Enhancing the CDI producer". And this sentence is included: `One of the advantages of using a CDI producer is that it can be tailored to your needs. For improved security, the createMongo method can be enhanced to include authentication with a user name and an encoded password`:

I think some version of this sentence, and a new section (maybe "Secure your CDI producer" or similar) would add some context keep the examples from running together, making them easier to parse


- [x] `You need to enable the Password Utilities feature in the server.xml file:`

Why? Does this apply to using a password or to adding TLS? It's not clear
**Made updates per our discussion for password utilities, TLS and CDI.**

- [x] `To access CouchDB with CDI, you can replace the MongoClient in the example...`
--->
`To access CouchDB with CDI, you can replace the MongoClient in the previous examples...`

- [x] `add the following to your CDI producer:`
--->
`add the following code to your CDI producer:`

- [x] `For more information, see Access any version of MongoDB with Open Liberty using CDI.`

recommend removing this sentence and adding the link to the "See also" section as `Blog: Access any version of MongoDB with Open Liberty using CDI`

**Comment by ManasiGandhi** [comment]
@chirp1 Here's the link to the updated draft https://draft-openlibertyio.mybluemix.net/docs/21.0.0.9/access-nosql-databases.html. 

- [x]  Previously I said "without enabling any feature" doesn't seem quite right. I also said to review the resources that Laura listed and reword/correct. I see that you removed the phrase instead of rewording/correcting it. I think, though, that a corrected phrase will add value to the short description and help add the essence of what the topic is about. To me, one way to interpret "You can access a NoSQL database with Open Liberty without enabling any feature in the server.xml file." is that the server.xml files does not contain a single feature in it. What isn't needed? Some features. To add clarity to the short description, figure out what those features are and include them in the short description in a meaningful way. **I updated the short description based on the problem statement for the topic and user pain points. Per Mark's suggestion, I removed any mention of features and added those details in the second paragraph**

 

- [x]    The opening paragraph now mentions CDI, but the new sentence repeats the title of the topic. The short description documentation says not to repeat the title. (See info about writing short descriptions on the peer review cheat sheet link.) I think that when you get the features figured out that I mentioned in the previous bullet, you can combine that info with the CDI reference to have a more meaningful short description in two sentences and about 50 words. **Updated in the new draft as compared to the working draft.**

- [x]  I mentioned that the "You can configure variables like the user and password " wording and the "To add configuration" wording are confusing, oddly worded. I asked that you rewrite the information because the info that you were trying to convey is meaningful to the user because it would describe some of what is going on in the code. However, I see that you removed the "You can configure variables like the user and password "wording. I suggest that you add back in wording about the variables to give more context to the user. Your "You can configure variables like the user and password " wording seems to indicate that "user" and "password" are variables, but I don't think they are. **Updated the information based on user journey, where the user creates a CDI producer, uses the created CDI producer to inject the MongoDatabase instance into the application, adds security to the CDI producer by including authentication, and then configures the MongoDB driver with MicroProfile Config and adds code to server.env to pull the values for user and encodedPass.**

**Comment by dmuelle** [comment]
Hi Manasi, the updates look good and I think mapping the user journey and pain points as guiding principles is a good strategy. I have a few suggestions for clarity and to hopefully improve the flow of the topic across the user journey. 

- `A CDI producer can be modified according to your needs and you can configure the database driver by using MicroProfile Config.`

This sentence seems to be running two orthogonally related tasks/sections together into one. If you added a little more detail, it might provide a clearer overview of the subsequent content. Something like:

`You can modify a CDI producer according to your needs. For example, you can add authentication or enable transport layer security (TLS) for your database connection. You can also use MicroProfile Config to configure the database driver.`

- `Using a CDI producer eliminates the need to enable features like the MongoDB feature for MongoDB or the CouchDB feature for CouchDB. The MongoDB and CouchDB features are stabilized after Open Liberty 19.0.0.6.`
reduce wordiness here and use the proper text name for features:
`Using a CDI producer eliminates the need to enable features like the MongoDB Integration feature or the CouchDB Integration feature. These features are stabilized after Open Liberty 19.0.0.6.`
Also, check the embedded links for the features, i think they are reversed.

- `You can customize a CDI producer according to your needs. The createMongo method can be enhanced to include authentication for additional security. You need to enable the Password Utilities feature in the server.xml file to include authentication with a username and an encoded password:`

I don't think you need to repeat the first sentence here as it is already in the intro
The "You need to enable password utilities..." sentence makes it sound like this is what the example is showing. Also makes it sound like this is server.xml code, which it is not. I think you should add another sentence to intro the example, something like
`You need to enable the Password Utilities feature in the server.xml file to include authentication in your CDI producer. 
In the following example, the CDI producer is enhanced to provide a username and encoded password for the database connection.`

- `You must enable TLS for your database connection. Add the JSSEHelper class to the createMongo method to get the SSLContext class, as shown in the following example:`

This makes it sound like the example shows how to enable TLS for the connection, and is also unclear whether  TLS and authentication are interdependent, which they are not (in this case at least). It might be more clear to say something like
`If you want to connect over TLS, you must first [configure TLS for your Open Liberty server](https://openliberty.io/docs/21.0.0.8/reference/feature/transportSecurity-1.0.html#_configure_transport_layer_security_tls). Then, you can modify your CDI producer to add the JSSEHelper class to the createMongo method and get the SSLContext class, as shown in the following example:`

- `Add the following snippet to your microprofile-config.properties or server.env file, to pull the values for user and encodedPass into the MongoProducer class`
I don't think you need the comma after `file`

- `user and encodedPass`---> ` the username and encoded password`
- `You can configure access to any version of CouchDB with a CDI producer in Open Liberty. To access CouchDB with CDI, you can replace the MongoClient in the previous examples with any CouchDB client or the Cloudant service.`

i think sentence should move up to the end of the "Access MongoDB with CDI" section, just before "securing..."

- you might add Guide: [Injecting dependencies into microservices](https://openliberty.io/guides/cdi-intro.html) to the see also section

**Comment by ManasiGandhi** [comment]
@dmuelle I worked on your review. Here's the updated link https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.9/access-nosql-databases.html

**Comment by chirp1** [comment]
@ManasiGandhi @dmuelle Hi Manasi, we discussed your topic on the 1PM Content review call today. 
The updates that you made based on the discussion on the call look good to me. The items from the call that you made updates for are as follows.
- The two sentence introductory paragraph needs mongoDB and cloudDB added back in, but indicating that they are examples. 
- The stabilization information for mongoDB and cloudDB needs to be corrected so that it indicates the correct version for each database and so that the stabilization starts with those versions, not some version after those versions.
- "...values for user and encodedPass into...". Correct so that a noun is after the terms or you use English wording.
- Information in a MongoDB section said that to use couchDB with CDI, replace mongodb in the examples with couchdb. However, a user might not spot this information in a section that had only the mongoDB reference in the title. Move the couchDB reference outside a MongoDB section.

**Comment by dmuelle** [comment]
LGTM, thanks!

**Comment by ManasiGandhi** [comment]
https://docs-staging-openlibertyio.mybluemix.net/docs/21.0.0.9/access-nosql-databases.html
